File: app/src/main/java/com/salino/sali/data/repository/HouseholdRepositoryImpl.kt
package com.salino.sali.data.repository

import com.google.firebase.firestore.ListenerRegistration
import com.salino.sali.data.local.source.HouseholdLocalDataSource
import com.salino.sali.data.model.Household
import com.salino.sali.data.model.HouseholdMember
import com.salino.sali.data.remote.source.HouseholdRemoteDataSource
import com.salino.sali.di.IoDispatcher
import com.salino.sali.domain.repository.AuthRepository
import com.salino.sali.domain.repository.HouseholdRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HouseholdRepositoryImpl @Inject constructor(
    private val remoteDataSource: HouseholdRemoteDataSource,
    private val localDataSource: HouseholdLocalDataSource,
    private val authRepository: AuthRepository,
    @IoDispatcher ioDispatcher: CoroutineDispatcher
) : HouseholdRepository {

    private val scope = CoroutineScope(SupervisorJob() + ioDispatcher)
    private val householdListeners = mutableMapOf<String, ListenerRegistration>()
    private val memberListeners = mutableMapOf<String, ListenerRegistration>()

    override suspend fun createHousehold(name: String): Result<Household> = runCatching {
        val user = authRepository.observeCurrentUser().first()
            ?: throw IllegalStateException("Not signed in")
        val household = remoteDataSource.createHousehold(
            userId = user.id,
            displayName = user.displayName,
            name = name
        )
        localDataSource.cacheHousehold(household)
        household
    }

    override suspend fun joinHousehold(inviteCode: String): Result<Household> = runCatching {
        val user = authRepository.observeCurrentUser().first()
            ?: throw IllegalStateException("Not signed in")
        val household = remoteDataSource.joinHousehold(
            userId = user.id,
            displayName = user.displayName,
            inviteCode = inviteCode
        )
        localDataSource.cacheHousehold(household)
        household
    }

    override suspend fun getHousehold(householdId: String): Result<Household> = runCatching {
        val local = localDataSource.getHousehold(householdId)
        if (local != null) {
            refreshHousehold(householdId)
            return@runCatching local
        }

        val remote = remoteDataSource.getHousehold(householdId)
        localDataSource.cacheHousehold(remote)
        remote
    }

    override fun observeHousehold(householdId: String): Flow<Household?> {
        ensureHouseholdSync(householdId)
        return localDataSource.observeHousehold(householdId)
    }

    override fun observeHouseholdMembers(householdId: String): Flow<List<HouseholdMember>> {
        ensureMembersSync(householdId)
        return localDataSource.observeMembers(householdId)
    }

    override suspend fun getInviteCode(householdId: String): Result<String> = runCatching {
        getHousehold(householdId).getOrThrow().inviteCode
    }

    override suspend fun updateHouseholdName(householdId: String, newName: String): Result<Unit> = runCatching {
        remoteDataSource.updateHouseholdName(householdId, newName)
    }

    override suspend fun leaveHousehold(householdId: String): Result<Unit> = runCatching {
        val user = authRepository.observeCurrentUser().first()
            ?: throw IllegalStateException("Not signed in")
        householdListeners[householdId]?.remove()
        householdListeners.remove(householdId)
        memberListeners[householdId]?.remove()
        memberListeners.remove(householdId)
        remoteDataSource.leaveHousehold(householdId, user.id)
        localDataSource.clearHouseholdData(householdId)
    }

    override fun clearListeners() {
        householdListeners.values.forEach { it.remove() }
        householdListeners.clear()
        memberListeners.values.forEach { it.remove() }
        memberListeners.clear()
    }

    private fun ensureHouseholdSync(householdId: String) {
        if (householdListeners.containsKey(householdId)) return
        householdListeners[householdId] = remoteDataSource.listenToHousehold(householdId) { household ->
            scope.launch {
                localDataSource.cacheHousehold(household)
            }
        }
    }

    private fun ensureMembersSync(householdId: String) {
        if (memberListeners.containsKey(householdId)) return
        memberListeners[householdId] = remoteDataSource.listenToMembers(householdId) { members ->
            scope.launch {
                localDataSource.replaceMembers(householdId, members)
            }
        }
    }

    private fun refreshHousehold(householdId: String) {
        scope.launch {
            runCatching {
                val household = remoteDataSource.getHousehold(householdId)
                localDataSource.cacheHousehold(household)
            }
        }
    }
}

File: ios/Data/Repositories/HouseholdRepositoryImpl.swift
import Foundation
import FirebaseFirestore

@MainActor
final class HouseholdRepositoryImpl: HouseholdRepository {
    private let remoteDataSource: HouseholdRemoteDataSource
    private let localStore: LocalStore
    private let authRepository: AuthRepository
    private var householdListeners: [String: ListenerRegistration] = [:]
    private var memberListeners: [String: ListenerRegistration] = [:]

    init(remoteDataSource: HouseholdRemoteDataSource, localStore: LocalStore, authRepository: AuthRepository) {
        self.remoteDataSource = remoteDataSource
        self.localStore = localStore
        self.authRepository = authRepository
    }

    func createHousehold(name: String) async throws -> Household {
        guard let user = await firstCurrentUser() else { throw FirebaseConfigurationError.missingCurrentUser }
        let household = try await remoteDataSource.createHousehold(userId: user.id, displayName: user.displayName, name: name)
        localStore.cacheHousehold(household)
        return household
    }

    func joinHousehold(inviteCode: String) async throws -> Household {
        guard let user = await firstCurrentUser() else { throw FirebaseConfigurationError.missingCurrentUser }
        let household = try await remoteDataSource.joinHousehold(userId: user.id, displayName: user.displayName, inviteCode: inviteCode)
        localStore.cacheHousehold(household)
        return household
    }

    func getHousehold(householdId: String) async throws -> Household {
        if let local = localStore.fetchHousehold(householdId: householdId) {
            Task { @MainActor in
                if let remote = try? await remoteDataSource.getHousehold(householdId: householdId) {
                    localStore.cacheHousehold(remote)
                }
            }
            return local
        }
        let remote = try await remoteDataSource.getHousehold(householdId: householdId)
        localStore.cacheHousehold(remote)
        return remote
    }

    func observeHousehold(householdId: String) -> AsyncStream<Household?> {
        ensureHouseholdSync(householdId)
        return localStore.observeHousehold(householdId: householdId)
    }

    func observeHouseholdMembers(householdId: String) -> AsyncStream<[HouseholdMember]> {
        ensureMembersSync(householdId)
        return localStore.observeMembers(householdId: householdId)
    }

    func getInviteCode(householdId: String) async throws -> String {
        try await getHousehold(householdId: householdId).inviteCode
    }

    func updateHouseholdName(householdId: String, newName: String) async throws {
        try await remoteDataSource.updateHouseholdName(householdId: householdId, newName: newName)
    }

    func leaveHousehold(householdId: String) async throws {
        guard let user = await firstCurrentUser() else { throw FirebaseConfigurationError.missingCurrentUser }
        householdListeners.removeValue(forKey: householdId)?.remove()
        memberListeners.removeValue(forKey: householdId)?.remove()
        try await remoteDataSource.leaveHousehold(householdId: householdId, userId: user.id)
        localStore.clearHouseholdData(householdId: householdId)
    }

    func clearListeners() {
        householdListeners.values.forEach { $0.remove() }
        householdListeners.removeAll()
        memberListeners.values.forEach { $0.remove() }
        memberListeners.removeAll()
    }

    private func ensureHouseholdSync(_ householdId: String) {
        guard householdListeners[householdId] == nil else { return }
        householdListeners[householdId] = remoteDataSource.listenToHousehold(householdId: householdId) { [weak self] household in
            Task { @MainActor in self?.localStore.cacheHousehold(household) }
        }
    }

    private func ensureMembersSync(_ householdId: String) {
        guard memberListeners[householdId] == nil else { return }
        memberListeners[householdId] = remoteDataSource.listenToMembers(householdId: householdId) { [weak self] members in
            Task { @MainActor in self?.localStore.replaceMembers(householdId: householdId, members: members) }
        }
    }

    private func firstCurrentUser() async -> UserProfile? {
        for await user in authRepository.observeCurrentUser() {
            return user
        }
        return nil
    }
}
